undofx/src/main/java/org/fxmisc/undo/impl/UndoManagerImpl.java

```diff
@@ -209,8 +229,10 @@ private void addChange(C change) {
     @SuppressWarnings("unchecked")
     private C[] merge(C c1, C c2) {
         Optional<C> merged = merge.apply(c1, c2);
-        if(merged.isPresent()) {
-            return (C[]) new Object[] { merged.get() };
+        if(merged.isPresent() && isNotIdentity.test(merged.get())) {
+            return isNotIdentity.test(merged.get())
```

Comment:
I've fixed this now on the PR. However, I haven't addressed the sloppy `canMerge` part as well on my local version because the test I wrote for it failed. I'll have to try later when I have more time.